### ISSUE TITLE
Retraining Zipvoice with BigVGAN 44khz as vocoder

### ISSUE THREAD

**Issue by moewiee** [k2-fsa/ZipVoice#16]
Hi, I want to try improving model performance by using BigVGAN as vocoder instead of Vocos. Can I just change the params in compute fbank then I can retrain the model without issues, or anything else I have to notice?

When I use BigVGAN 24khz 100 mels instead of Vocos, the output sounds almost like only noise.

**Comment by moewiee** [comment]
@zhu-han Thank you for your fast response!

**Comment by moewiee** [comment]
<img width="1091" alt="Image" src="https://github.com/user-attachments/assets/a0e92441-53c3-4488-8c45-b193db237510" />

The waveform and the spectrogram now look like this. I can still hear the text I want to generate, it's just buried in a lot of noise.

**Comment by zhu-han** [comment]
Yes, you'll need to re-train or fine-tune ZipVoice on BigVGAN features if you want to use BigVGAN vocoder.  

`tools/feature.py` is a soft link of `zipvoice/feature.py`. The `MelSpectrogramFeatures` in `zipvoice_infer.py` is redundant. I'll remove it later. Thanks for pointing it out.

> [@zhu-han](https://github.com/zhu-han) I have change the feature extraction method and the voice sound a lot clearer. However, it is still very noisy. Is that because the model is pretrained with the Vocos feature extractor so it can not be easily switch to BigVGAN feature extractor without re-train?
> 
> Also, I can find 3 MelSpectrogramFeatures classes in 3 different files: tools/feature.py zipvoice/feature.py and zipvoice_infer.py Can you please explain why there are definition of this class in different places? Thank you!

**Comment by wincing2** [comment]
> Yes, you'll need to re-train or fine-tune ZipVoice on BigVGAN features if you want to use BigVGAN vocoder.
> 
> `tools/feature.py` is a soft link of `zipvoice/feature.py`. The `MelSpectrogramFeatures` in `zipvoice_infer.py` is redundant. I'll remove it later. Thanks for pointing it out.
> 
> > [@zhu-han](https://github.com/zhu-han) I have change the feature extraction method and the voice sound a lot clearer. However, it is still very noisy. Is that because the model is pretrained with the Vocos feature extractor so it can not be easily switch to BigVGAN feature extractor without re-train?
> > Also, I can find 3 MelSpectrogramFeatures classes in 3 different files: tools/feature.py zipvoice/feature.py and zipvoice_infer.py Can you please explain why there are definition of this class in different places? Thank you!

Are you sure fine-tuning ZipVoice Base on BigVGAN features would work, since ZipVoice Base was pretrained on different features?

**Comment by jingzhaoou** [comment]
@moewiee  Any luck? I would like to use ZipVoice to generate voice with even better quality as well.

**Comment by yaguanghu** [comment]
I retrained the ZipVoice model based on BigVGAN on the Libritts dataset from scratch and have trained it to 35 epochs. The synthesized audio content can clearly be heard, but the sound quality is very poor. @zhu-han Do you have any suggestions?

The text corresponding to the sample audio is
`But the more forgetfulness had then prevailed, the more powerful was the force of remembrance when she awoke.`

[4992-23283-0000_vocos.wav](https://github.com/user-attachments/files/21884424/4992-23283-0000_vocos.wav)

[4992-23283-0000_bigvgan.wav](https://github.com/user-attachments/files/21884427/4992-23283-0000_bigvgan.wav)

**Comment by yaguanghu** [comment]
> > I retrained the ZipVoice model based on BigVGAN on the Libritts dataset from scratch and have trained it to 35 epochs. The synthesized audio content can clearly be heard, but the sound quality is very poor. [@zhu-han](https://github.com/zhu-han) Do you have any suggestions?
> > The text corresponding to the sample audio is `But the more forgetfulness had then prevailed, the more powerful was the force of remembrance when she awoke.`
> > [4992-23283-0000_vocos.wav](https://github.com/user-attachments/files/21884424/4992-23283-0000_vocos.wav)
> > [4992-23283-0000_bigvgan.wav](https://github.com/user-attachments/files/21884427/4992-23283-0000_bigvgan.wav)
> 
> One of our team members also tried to train a model using BigVGAN features, but the audio quality is not good. We don't a good solution for this yet.

Have you encountered similar problems? I tested WER on the librispeech-pc test set and the results are comparable with vocos vocoder, but the sound quality is very poor.

**Comment by zhu-han** [comment]
> > > I retrained the ZipVoice model based on BigVGAN on the Libritts dataset from scratch and have trained it to 35 epochs. The synthesized audio content can clearly be heard, but the sound quality is very poor. [@zhu-han](https://github.com/zhu-han) Do you have any suggestions?
> > > The text corresponding to the sample audio is `But the more forgetfulness had then prevailed, the more powerful was the force of remembrance when she awoke.`
> > > [4992-23283-0000_vocos.wav](https://github.com/user-attachments/files/21884424/4992-23283-0000_vocos.wav)
> > > [4992-23283-0000_bigvgan.wav](https://github.com/user-attachments/files/21884427/4992-23283-0000_bigvgan.wav)
> > 
> > 
> > One of our team members also tried to train a model using BigVGAN features, but the audio quality is not good. We don't a good solution for this yet.
> 
> Have you encountered similar problems? I tested WER on the librispeech-pc test set and the results are comparable with vocos vocoder, but the sound quality is very poor.

Yes, the audio quality is lower than Vocos, though it sounds better than your example.

**Comment by yaguanghu** [comment]
> > > > I retrained the ZipVoice model based on BigVGAN on the Libritts dataset from scratch and have trained it to 35 epochs. The synthesized audio content can clearly be heard, but the sound quality is very poor. [@zhu-han](https://github.com/zhu-han) Do you have any suggestions?
> > > > The text corresponding to the sample audio is `But the more forgetfulness had then prevailed, the more powerful was the force of remembrance when she awoke.`
> > > > [4992-23283-0000_vocos.wav](https://github.com/user-attachments/files/21884424/4992-23283-0000_vocos.wav)
> > > > [4992-23283-0000_bigvgan.wav](https://github.com/user-attachments/files/21884427/4992-23283-0000_bigvgan.wav)
> > > 
> > > 
> > > One of our team members also tried to train a model using BigVGAN features, but the audio quality is not good. We don't a good solution for this yet.
> > 
> > 
> > Have you encountered similar problems? I tested WER on the librispeech-pc test set and the results are comparable with vocos vocoder, but the sound quality is very poor.
> 
> Yes, the audio quality is lower than Vocos, though it sounds better than your example.

@zhu-han I think my LibriTTS experiment on HiFTNet vocoder is working fine. Would you welcome a PR for this?

**Comment by zhu-han** [comment]
Is HiFTNet better than Vocos?

**Comment by yaguanghu** [comment]
> Is HiFTNet better than Vocos?

The sound quality sounds better to me. but strangely, the utmos score actually dropped.
Perhaps because I haven't adjusted any hyperparameters, I'm currently using the same parameters as Vocos.

**Comment by zhu-han** [comment]
> > Is HiFTNet better than Vocos?
> 
> The sound quality sounds better to me. but strangely, the utmos score actually dropped. Perhaps because I haven't adjusted any hyperparameters, I'm currently using the same parameters as Vocos.

Can you provide some samples for this comparison?

**Comment by yaguanghu** [comment]
> > > Is HiFTNet better than Vocos?
> > 
> > 
> > The sound quality sounds better to me. but strangely, the utmos score actually dropped. Perhaps because I haven't adjusted any hyperparameters, I'm currently using the same parameters as Vocos.
> 
> Can you provide some samples for this comparison?

Sure. Here is the an example from LibiriSpeech PC testset. If you listen carefully, you’ll notice there’s some background electronic reverberation in the audio from Vocos.

[1188-133604-0001_HiFTNet.wav](https://github.com/user-attachments/files/23413303/1188-133604-0001_HiFTNet.wav)
[1188-133604-0001_vocos.wav](https://github.com/user-attachments/files/23413302/1188-133604-0001_vocos.wav)

**Comment by zhu-han** [comment]
Thanks for sharing this. I think the HiFTNet audio also has a specific artifact pattern that is not present in the Vocos audio. It's hard to say whether the HiFTNet audio has better audio quality. Additionally, HiFTNet is much slower than Vocos (according to the paper). I think there may not be a clear benefit to integrating this vocoder into the master branch for now.

**Comment by yaguanghu** [comment]
> Thanks for sharing this. I think the HiFTNet audio also has a specific artifact pattern that is not present in the Vocos audio. It's hard to say whether the HiFTNet audio has better audio quality. Additionally, HiFTNet is much slower than Vocos (according to the paper). I think there may not be a clear benefit to integrating this vocoder into the master branch for now.

OK, I see. I made this change mainly to solve the reverberation problem in multilingual synthesized audios, such as French, German, Spanish, etc. If anyone is interested, feel free to contact me.
Thank you for your reply.

**Comment by zgxiangyang** [comment]
> > > > Is HiFTNet better than Vocos?
> > > 
> > > 
> > > The sound quality sounds better to me. but strangely, the utmos score actually dropped. Perhaps because I haven't adjusted any hyperparameters, I'm currently using the same parameters as Vocos.
> > 
> > 
> > Can you provide some samples for this comparison?
> 
> Sure. Here is the an example from LibiriSpeech PC testset. If you listen carefully, you’ll notice there’s some background electronic reverberation in the audio from Vocos.
> 
> [1188-133604-0001_HiFTNet.wav](https://github.com/user-attachments/files/23413303/1188-133604-0001_HiFTNet.wav) [1188-133604-0001_vocos.wav](https://github.com/user-attachments/files/23413302/1188-133604-0001_vocos.wav)

@yaguanghu I think the sound quality of the HiFTNet version is better, could you please share the prompt audio.

**Comment by Lokesh-Kumar1998** [comment]
> > Thanks for sharing this. I think the HiFTNet audio also has a specific artifact pattern that is not present in the Vocos audio. It's hard to say whether the HiFTNet audio has better audio quality. Additionally, HiFTNet is much slower than Vocos (according to the paper). I think there may not be a clear benefit to integrating this vocoder into the master branch for now.
> 
> OK, I see. I made this change mainly to solve the reverberation problem in multilingual synthesized audios, such as French, German, Spanish, etc. If anyone is interested, feel free to contact me. Thank you for your reply.

Hi, need some help for hiftnet integration in vocos.. did you make make changes in filterbank extraction code only?

**Comment by yaguanghu** [comment]
> > > > > Is HiFTNet better than Vocos?
> > > > 
> > > > 
> > > > The sound quality sounds better to me. but strangely, the utmos score actually dropped. Perhaps because I haven't adjusted any hyperparameters, I'm currently using the same parameters as Vocos.
> > > 
> > > 
> > > Can you provide some samples for this comparison?
> > 
> > 
> > Sure. Here is the an example from LibiriSpeech PC testset. If you listen carefully, you’ll notice there’s some background electronic reverberation in the audio from Vocos.
> > [1188-133604-0001_HiFTNet.wav](https://github.com/user-attachments/files/23413303/1188-133604-0001_HiFTNet.wav) [1188-133604-0001_vocos.wav](https://github.com/user-attachments/files/23413302/1188-133604-0001_vocos.wav)
> 
> [@yaguanghu](https://github.com/yaguanghu) I think the sound quality of the HiFTNet version is better, could you please share the prompt audio.

The prompt audio is from LibiriSpeech PC testset, you can find it in the test set.

**Comment by yaguanghu** [comment]
> > > Thanks for sharing this. I think the HiFTNet audio also has a specific artifact pattern that is not present in the Vocos audio. It's hard to say whether the HiFTNet audio has better audio quality. Additionally, HiFTNet is much slower than Vocos (according to the paper). I think there may not be a clear benefit to integrating this vocoder into the master branch for now.
> > 
> > 
> > OK, I see. I made this change mainly to solve the reverberation problem in multilingual synthesized audios, such as French, German, Spanish, etc. If anyone is interested, feel free to contact me. Thank you for your reply.
> 
> Hi, need some help for hiftnet integration in vocos.. did you make make changes in filterbank extraction code only?

Yes. What you need to do is change the way of feature extraction and then retrain the model. In my experiments, I didn't change any of the training hyperparameters.

**Comment by nshmyrev** [comment]
If you look to improve quality its better to predict VAE latents instead of mel. Like wavvae from MegaTTS3 or encodec latents from pflow_encodec. BigVGAN is not very relevant.

**Comment by Lokesh-Kumar1998** [comment]
> > > > Thanks for sharing this. I think the HiFTNet audio also has a specific artifact pattern that is not present in the Vocos audio. It's hard to say whether the HiFTNet audio has better audio quality. Additionally, HiFTNet is much slower than Vocos (according to the paper). I think there may not be a clear benefit to integrating this vocoder into the master branch for now.
> > > 
> > > 
> > > OK, I see. I made this change mainly to solve the reverberation problem in multilingual synthesized audios, such as French, German, Spanish, etc. If anyone is interested, feel free to contact me. Thank you for your reply.
> > 
> > 
> > Hi, need some help for hiftnet integration in vocos.. did you make make changes in filterbank extraction code only?
> 
> Yes. What you need to do is change the way of feature extraction and then retrain the model. In my experiments, I didn't change any of the training hyperparameters.

Hi, i tried same with hifigan but got issue of blank audios, can I contact you regarding that ?

**Comment by wincing2** [comment]
> If you look to improve quality its better to predict VAE latents instead of mel. Like wavvae from MegaTTS3 or encodec latents from pflow_encodec. BigVGAN is not very relevant.

I've tried other features like VAE latents or encoder latents with matcha-tts. They didn't work well. The synthesized speech is audible but not good. It seems to be that the flow matching algorithm in matcha-tts only works with features with "locality" (nearby features has similiar values, as in mel-spectrograms and other 'real' images). Haven't tried latent with zipvoice because it would require train from scratch which would be a lot of compute.

**Comment by Lokesh-Kumar1998** [comment]
> > > > Thanks for sharing this. I think the HiFTNet audio also has a specific artifact pattern that is not present in the Vocos audio. It's hard to say whether the HiFTNet audio has better audio quality. Additionally, HiFTNet is much slower than Vocos (according to the paper). I think there may not be a clear benefit to integrating this vocoder into the master branch for now.
> > > 
> > > 
> > > OK, I see. I made this change mainly to solve the reverberation problem in multilingual synthesized audios, such as French, German, Spanish, etc. If anyone is interested, feel free to contact me. Thank you for your reply.
> > 
> > 
> > Hi, need some help for hiftnet integration in vocos.. did you make make changes in filterbank extraction code only?
> 
> Yes. What you need to do is change the way of feature extraction and then retrain the model. In my experiments, I didn't change any of the training hyperparameters.

I have feature.py file in utils folder: 
from dataclasses import dataclass
from typing import Union

import numpy as np
import torch
import torch.nn.functional as F

from lhotse.features.base import FeatureExtractor, register_extractor
from lhotse.utils import Seconds, compute_num_frames

from librosa.filters import mel as librosa_mel_fn


@dataclass
class HiftNetFbankConfig:
    sampling_rate: int = 22050
    n_fft: int = 1024
    hop_length: int = 256
    win_length: int = 1024
    n_mels: int = 80
    fmin: int = 0
    fmax: int = 8000


@register_extractor
class HiftNetFbank(FeatureExtractor):
    """
    Bit-matched HiftNet mel extractor for ZipVoice (Lhotse-compatible).
    """

    name = "HiftNetFbank"
    config_type = HiftNetFbankConfig

    def __init__(self, num_channels: int = 1):
        super().__init__(config=HiftNetFbankConfig())
        assert num_channels in (1, 2)
        self.num_channels = num_channels

        # NOTE: FeatureExtractor is NOT nn.Module → no register_buffer
        self.mel_basis = torch.from_numpy(
            librosa_mel_fn(
                sr=self.config.sampling_rate,
                n_fft=self.config.n_fft,
                n_mels=self.config.n_mels,
                fmin=self.config.fmin,
                fmax=self.config.fmax,
            )
        ).float()

        self.hann_window = torch.hann_window(self.config.win_length)

    def _stft(self, y: torch.Tensor) -> torch.Tensor:
        pad = (self.config.n_fft - self.config.hop_length) // 2
        y = F.pad(y, (pad, pad), mode="reflect")

        spec = torch.stft(
            y,
            n_fft=self.config.n_fft,
            hop_length=self.config.hop_length,
            win_length=self.config.win_length,
            window=self.hann_window.to(y.device),
            center=False,
            return_complex=True,
        )

        # EXACT HiftNet magnitude
        spec = torch.view_as_real(spec)
        spec = torch.sqrt(spec.pow(2).sum(-1) + 1e-9)
        return spec

    def extract(self, samples, sampling_rate):
        assert sampling_rate == self.config.sampling_rate

        is_numpy = False
        if not isinstance(samples, torch.Tensor):
            samples = torch.from_numpy(samples)
            is_numpy = True

        if samples.dim() == 1:
            samples = samples.unsqueeze(0)

        if self.num_channels == 1 and samples.size(0) == 2:
            samples = samples.mean(dim=0, keepdim=True)

        samples = samples.to(self.mel_basis.device)

        spec = self._stft(samples)
        mel = torch.matmul(self.mel_basis.to(spec.device), spec)
        mel = torch.log(torch.clamp(mel, min=1e-5))

        mel = mel.squeeze(0).transpose(0, 1)

        expected_num_frames = compute_num_frames(
            samples.shape[1] / sampling_rate,
            self.frame_shift,
            sampling_rate,
        )

        cur_num_frames = mel.size(0)

        if cur_num_frames > expected_num_frames:
            mel = mel[:expected_num_frames]
        elif cur_num_frames < expected_num_frames:
            pad_frames = expected_num_frames - cur_num_frames
            last_frame = mel[-1:].repeat(pad_frames, 1)
            mel = torch.cat([mel, last_frame], dim=0)


        return mel.cpu().numpy() if is_numpy else mel


    def feature_dim(self, sampling_rate: int) -> int:
        return self.config.n_mels

    @property
    def frame_shift(self) -> Seconds:
        return self.config.hop_length / self.config.sampling_rate
